ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Protocol/Praos.hs

```diff
@@ -78,31 +78,97 @@ import           Ouroboros.Consensus.Protocol.Abstract
+--
+--      > (ρ_y, ρ_π) ≜ VRF_evalProve(η, slⱼ, NONCE) see Fig 9
+--
+-- - σⱼ: a signature on (st, d, slⱼ, B_πj, ρ) with the signing key
+--       for the slot slⱼ for the stakeholder Uᵢ.
+--
+-- Protocol parameters:
+-- - k: maximum number of blocks we can rollback.
+-- - R: number of slots per epoch.
+-- - f: the active slots coefficient, specifies roughly the proportion of
+--      occupied slots per epoch.
+--
+-- Some values you will encounter:
+-- - η: The epoch's nonce. Captures entropy from the block chain. See Fig 8 in
+--      praos paper for where it is used and Fig 10 for how it is defined.
+--      Defined as the hash of the η from the previous epoch, this epoch number
+--      and the ρ of the first 2/3 of the blocks in the previous epoch.
+--
+--      > η_e ≜ HASH(η_{e-1} || e || ρ_{e-1,0} ... ρ_{e-1, 2R/3})
+--
+-- - Tᵢ: the leader threshold for a specific stakeholder considering its
+--       relative stake (therefore depending on the slot). Defined in the Praos
+--       paper in Figure 4 using the definition for ϕ_f(αᵢ) from section 3.3.
+--
+--       > Tᵢ ≜ 2^(l_VRF) * pᵢ
+--       > pᵢ = ϕ_f(αᵢ) ≜ 1 - (1 - f)^(αᵢ)
+
 {-------------------------------------------------------------------------------
-  Fields required by Praos in the header
+  Fields required by Praos in the header of the block
 -------------------------------------------------------------------------------}
 
 -- | The fields that Praos required in the header
-data PraosFields c toSign = PraosFields {
-      praosSignature   :: SignedKES (PraosKES c) toSign
-    , praosExtraFields :: PraosExtraFields c
+data PraosFields crypto typeBeingSigned = PraosFields {
+      praosSignature   :: SignedKES (PraosKES crypto) typeBeingSigned
+      -- ^The signature of the header (σⱼ)
+    , praosExtraFields :: PraosExtraFields crypto
     }
   deriving (Generic)
 
 instance (PraosCrypto c, Typeable toSign) => NoThunks (PraosFields c toSign)
   -- use generic instance
 
+deriving instance PraosCrypto c => Show (PraosFields c toSign)
+deriving instance PraosCrypto c => Eq   (PraosFields c toSign)
+
 -- | Fields that should be included in the signature
 data PraosExtraFields c = PraosExtraFields {
       praosCreator :: CoreNodeId
+      -- ^Which node signed this block.
     , praosRho     :: CertifiedVRF (PraosVRF c) (Natural, SlotNo, VRFType)
+      -- ^The output and proof of ρ.
     , praosY       :: CertifiedVRF (PraosVRF c) (Natural, SlotNo, VRFType)
+      -- ^The output and proof of y.
     }
   deriving (Generic)
 
 instance PraosCrypto c => NoThunks (PraosExtraFields c)
   -- use generic instance
 
+deriving instance PraosCrypto c => Show (PraosExtraFields c)
+deriving instance PraosCrypto c => Eq   (PraosExtraFields c)
+
+-- | A validate view is an association from the (@signed@) value to the
+-- @PraosFields@ that contains the signature that sign it.
+--
+-- In this case, PraosCrypto forces @signed@ to be @(Natural, SlotNo, VRFType)@ (why?)
```

Comment:
Let's work through the `why?` question. I don't immediately know the answer -- we'll have to look some more.

Reply:
There are lots and layers to the crypto types. I have trouble keeping them in my brain.

In this case, poking through it for the `n`-th time, I think this is the story:

  * In a value `PraosValidateView x y`,  `x` contains a KES signature and `y` is the data that was signed with the KES key to create that signature.

  * So `signed` here will be the parts of the block that are signed to create the signature in the block header -- that's what we use KES signatures for. See `Ouroboros.Consensus.Mock.Ledger.Block.Praos.SignedSimplePraos`. The consensus algorithm doesn't care what that is, so we use the existential here. For this mock implementation, we could probably have just arranged to import `SignedSimplePraos` here, but that'd be less flexible and wouldn't respect the divide we have between consensus protocol and block types (protocols don't depend on block details directly).
  
  * The `(Natural, SlotNo, VRFType)` type, on the other hand, is what is signed by the VRF key, not the KES key. The protocol itself does that, which is why that type's "VRF signability" constraint is baked into the `PraosCrypto` superclass.

Reply:
I added a comment explaining what you just said above, that this could be simplified but from consensus point of view, it doesn't care what is the actual underlying type so we just use the existential. I think it is useful for someone who comes without prior knowledge in order to not puzzle around "there is only one type that fulfills this, why not simplify it?".